presto-main/src/main/java/com/facebook/presto/execution/PlanFlattener.java

```diff
@@ -0,0 +1,245 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.execution;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.analyzer.QueryExplainer;
+import com.facebook.presto.sql.planner.PartitioningHandle;
+import com.facebook.presto.sql.planner.PartitioningScheme;
+import com.facebook.presto.sql.planner.PlanFragment;
+import com.facebook.presto.sql.planner.SimplePlanVisitor;
+import com.facebook.presto.sql.planner.SubPlan;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.plan.PlanFragmentId;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.PlanNodeId;
+import com.facebook.presto.sql.planner.plan.RemoteSourceNode;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.databind.JsonMappingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.JsonSerializer;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializerProvider;
+import com.fasterxml.jackson.databind.jsontype.TypeSerializer;
+import com.fasterxml.jackson.databind.module.SimpleModule;
+import com.google.common.collect.ImmutableList;
+
+import javax.inject.Inject;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static java.util.Objects.requireNonNull;
+
+public class PlanFlattener
+{
+    private final QueryExplainer explainer;
+    private final FlatteningVisitor visitor;
+
+    @Inject
+    public PlanFlattener(
+            QueryExplainer explainer,
+            ObjectMapper objectMapper)
+    {
+        this.explainer = requireNonNull(explainer, "explainer is null");
+        this.visitor = new FlatteningVisitor(requireNonNull(objectMapper, "objectMapper is null"));
+    }
+
+    public FlattenedPlan flatten(SubPlan rootStage, Session session)
+    {
+        return new FlattenedPlan(rootStage.getAllFragments().stream()
+                .map(fragment -> FlattenedPlanFragment.fromPlanFragment(fragment, explainer, session, visitor))
+                .collect(Collectors.toList()));
+    }
+
+    public static class FlattenedPlan
+    {
+        private final List<FlattenedPlanFragment> fragments;
+
+        private FlattenedPlan(List<FlattenedPlanFragment> fragments)
+        {
+            this.fragments = ImmutableList.copyOf(requireNonNull(fragments, "fragments is null"));
+        }
+
+        @JsonProperty
+        public List<FlattenedPlanFragment> getFragments()
+        {
+            return fragments;
+        }
+    }
+
+    public static class FlattenedPlanFragment
+    {
+        private final String textPlan;
+        private final List<JsonNode> nodes;
```

Comment:
Instead of using Json node you may be able to use a Class that wraps a String with `@JsonValue` and `@JsonRawValue` annotations